src/sgame/sg_bot_util.cpp

```diff
@@ -1370,6 +1382,31 @@ float BotAimAngle( gentity_t *self, vec3_t pos )
 	return AngleBetweenVectors( forward, ideal );
 }
 
+// fills allies with players of a team, and return the number of entities found.
+// pre-condition:
+// * team is playable
+// * allies is an array of at least alliesSize unsigned integers
+// * allies memory is initialized
+// * there are less than MAX_CLIENTS players in team
+// * there is enough space to store all team's player's references into allies
+unsigned int FindAllies( unsigned int allies[], size_t alliesSize, team_t team )
+{
+	ASSERT( G_IsPlayableTeam( team ) );
+	gentity_t *testEntity;
+	unsigned int numAllies = 0;
+	for ( size_t i = 0; i < MAX_CLIENTS && numAllies < alliesSize; i++ )
```

Comment:
You could use `ForEntities<ClientComponent>` and then check the `TeamComponent`.

Also it seems like you wouldn't want to count yourself as an ally.

Reply:
> You could use ForEntities<ClientComponent> and then check the TeamComponent.

I'm reluctant to do that, for several reasons, some of those being:

* it is not documented, the only "documentation" in there is: `// Definitions of ForEntities //`
* it's a variadic template, without static_assert, which means it will end in me wasting lot of time debugging cryptic error messages, to replace a single, easy to read for anyone, line.
* from what I could see, it relies on a std::set. I'm not sure using a red-black tree to store (sorted...) pointers is a good idea, for a start, nor that it is a good idea to use that to iterate over a flat array. From what I understand by reading code, performance-wise, it would be a bad idea (this function is likely to be called several times per "server frame")

> Also it seems like you wouldn't want to count yourself as an ally.

Right, the name is probably not nicely chosen. I've renamed it as ListTeamMembers.

Reply:
> 
> 
> > You could use ForEntities and then check the TeamComponent.
> 
> I'm reluctant to do that, for several reasons, some of those being:
> 
>     * it is not documented, the only "documentation" in there is: `// Definitions of ForEntities //`
>     * it's a variadic template, without static_assert, which means it will end in me wasting lot of time debugging cryptic error messages, to replace a single, easy to read for anyone, line.

If you refuse to use an undocumented function, you're not going to get far in this codebase. It should be doable to figure out the usage from examples (which invariably have a single argument).

> 
>     * from what I could see, it relies on a std::set. I'm not sure using a red-black tree to store (sorted...) pointers is a good idea, for a start, nor that it is a good idea to use that to iterate over a flat array. From what I understand by reading code, performance-wise, it would be a bad idea (this function is likely to be called several times per "server frame")

Since we are iterating over all entities, rather than looking one up, it is similar to iterating a linked list. Yes, there is room for performance improvements, but the penalty of a linked list over an array is not at  enough to justify spurning a clean solution for one that is, in your own words, "pure guessing and luck" :smile: 

> > Also it seems like you wouldn't want to count yourself as an ally.
> 
> Right, the name is probably not nicely chosen. I've renamed it as ListTeamMembers.
> I'm not sure it should be changed to skip self.

OK, but you still need to exclude the bot's own weapons somehow when doing the calculations.